libdrgn/program.c

```diff
@@ -872,30 +880,81 @@ drgn_thread_iterator_create(struct drgn_program *prog,
 			    struct drgn_thread_iterator **ret)
 {
 	struct drgn_error *err;
-	err = drgn_program_cache_threads(prog);
-	if (err)
-		return err;
-
-	*ret = malloc(sizeof(struct drgn_thread_iterator));
+	*ret = malloc(sizeof(**ret));
 	if (!*ret)
 		return &drgn_enomem;
-	(*ret)->iterator = drgn_thread_set_first(&prog->thread_set);
-	return NULL;
+	(*ret)->prog_is_live = prog->flags & DRGN_PROGRAM_IS_LIVE;
+	if ((*ret)->prog_is_live) {
+		if (!(prog->flags & DRGN_PROGRAM_IS_LINUX_KERNEL)) {
+			err = drgn_error_create(
+				DRGN_ERROR_INVALID_ARGUMENT,
+				"iterating over the threads of a live userspace process is not currently supported");
+			goto out;
+		}
+		struct drgn_object ns;
+		drgn_object_init(&ns, prog);
+		err = drgn_program_find_object(prog, "init_pid_ns", NULL,
+					       DRGN_FIND_OBJECT_VARIABLE, &ns);
+		if (err)
+			goto out;
+		err = drgn_object_address_of(&ns, &ns);
+		if (err)
+			goto out;
+		err = task_iter_init(prog, &ns, &(*ret)->task_iterator);
+		if (err)
+			goto out;
+	} else {
+		err = drgn_program_cache_threads(prog);
+		if (err)
+			goto out;
+		(*ret)->iterator = drgn_thread_set_first(&prog->thread_set);
+	}
+
+out:
+	if (err)
+		free(*ret);
+	return err;
 }
 
 LIBDRGN_PUBLIC void
 drgn_thread_iterator_destroy(struct drgn_thread_iterator *it)
 {
+	if (it->prog_is_live)
+		task_iter_deinit(&it->task_iterator);
 	free(it);
 }
 
 LIBDRGN_PUBLIC struct drgn_error *
 drgn_thread_iterator_next(struct drgn_thread_iterator *it,
 			  struct drgn_thread **ret)
 {
-	*ret = it->iterator.entry;
-	if (it->iterator.entry)
-		it->iterator = drgn_thread_set_next(it->iterator);
+	if (it->prog_is_live) {
+		task_iter_entry_t *entry;
+		struct drgn_error *err =
+			task_iter_next(&it->task_iterator, &entry);
+		if (err)
+			return err;
+		if (!entry) {
+			*ret = NULL;
+			return NULL;
+		}
+		it->entry.thread = *entry;
```

Comment:
Use `drgn_object_copy()` to copy objects rather than direct assignment. (Technically, direct assignment like this works for dumb values, but again I'd rather not assume that anywhere in the codebase in case we want to change that later.)